README.md
# HACKtheMACHINE: Data Science and the Seven Seas

[![EMAIL ADDRESS](https://img.shields.io/badge/CONTACT-EMAIL-brightgreen.svg)](mailto:danielrbrownjr+htm@gmail.com)
[![MIT LICENSE](https://img.shields.io/badge/LICENSE-MIT-bri.svg)](./LICENSE.md)
[![Linkedin](https://img.shields.io/badge/Linkedin-Daniel-red.svg?style=social)](http://tiny.cc/danielbrown)
[![Linkedin](https://img.shields.io/badge/Linkedin-Frederick-red.svg?style=social)](http://linkedin.com/in/frederick-miller-06020386)

This is the github repository for Team 6 "The Team Around the Corner" for [HACKtheMachine](https://www.hackthemachine.ai/): Data Science and the Seven Seas hosted at [MIT CSAIL](http://www.csail.mit.edu/) by the [Navy Postgraduate School](http://www.nps.edu/), [MIT](http://www.mit.edu) and [NAVSEA](http://www.navsea.navy.mil/).  

## Team Members:

Daniel Brown (Former USN MM1/SS(NUC)) SSBN-732, MPS Data Analytics student at Northeastern University 
Nicholas Rosato (Former USN MM2/SS(NUC)) SSN-22, SSN-690, Software Engineer for IBM   
Frederick Miller, NSWC Dahlgren  
Avory Bryant, NSWC Dahlgren  
Stu Bingham, NSWC Dahlgren  

![HACKtheMACHINE Image](https://static1.squarespace.com/static/596d24cd4402430bb863ffad/t/5977bc33ebbd1a2b573d2df7/1505600711911/?format=1500w)


CHANGELOG.md
# Changelog

[![Version 0.1](https://img.shields.io/badge/V%200.1-Daniel-blue.svg)](./template.R) Added first data processing program to project

[![Version 0.1](https://img.shields.io/badge/V%200.2-Daniel-blue.svg)](./template.py) Started coding python data processing program

[![Version 1.0](https://img.shields.io/badge/V%201.0-Team%206-green.svg)](./hackthemachine.R) Submitted completed code.


template.R
#### R Template for HACKtheMACHINE September 22nd and 23rd #####################

### Data Preprocessing #########################################################
## Installing required packages and calling libraries ##########################
install.packages("tidyverse")
install.packages("caTools")
install.packages("cluster")

library(magrittr) # adds pipe operator
library(tidyverse) # general data tidy package
library(caTools) # tools for data training
library(cluster) # adds hierarchical clustering 

## Importing dataset ###########################################################
dataset <- read.csv("dataset.csv") # imports data to data.frame

## Accounting for missing data #################################################
# Taking care of missing data (If there is no missing data, do not use)
dataset$data.column2 <- ifelse(is.na(dataset$data.column3),
                                ave(dataset$data.column3, 
                                    FUN = function(x) mean(x, na.rm = TRUE)),
                                dataset$data.column3)

## Scaling data features #######################################################
untouched <- select(dataset, data.column1)
dataset %<>% select(data.column2, data.column3, data.column4) # remove data we 
# don't want to scale
dataset %<>% scale # comment out if unnecessary
dataset %<>% cbind(unscaled, .) # return data to data.frame
rm(untouched) # removes unscaled vector

## Handling categorical data ###################################################
dataset$data.column1 %<>% factor(levels = c('Male', 'Female'), labels = c(0, 1))

## Splitting data into training and testing sets ###############################
set.seed(4) # ensures all parties get similar results
split <-dataset %>% sample.split(SplitRatio = 0.75)
data.train <- dataset %>% subset(split == TRUE)
data.test <- dataset %>% subset(split == FALSE)
rm(split)

### Data manipulation ##########################################################
## Creating linear regression based on one independent variable ################
y.pred <-
  lm(formula = data.column4 ~ data.column2, data = data.train ) %>% 
  predict(data.test)

### Data visualization #########################################################
## Creating a scatterplot based on some of the data ############################
ggplot(dataset, aes(data.column2, data.column4)) +
  geom_point(size = 4, color = "Red")


nanfilter.R
library(magrittr)
library(tidyverse)

dataset <- read.csv("classC_ship1_MRG.csv")

names(dataset)


MRG1 <-
  dataset$MRG2.TN.GEAR.FWD.BRG.T %>%
  is.na %>%
  which %>%
  dataset[.,]
x1 <-
  dataset %>%
  names %>%
  grep("MRG1", .)
x1 <- c(56, x1)
MRG1 <- MRG1[, x1]

MRG2 <-
  dataset$MRG1.TN.GEAR.FWD.BRG.T %>%
  is.na %>%
  which %>%
  dataset[.,]
x2 <-
  dataset %>%
  names %>%
  grep("MRG2", .)
x2 <- c(56, x2)
MRG2 <- MRG2[, x2]



colnames(MRG1) <- 
  c("time", "TN.GEAR.FWD.BRG.T",  "PRI.TN.GEAR.ENGGED", "PRI.TN.GEAR.DISENG", 
    "LO.PMP.A.DISC.PRES", "THR.BRG.AHEAD.TEMP", "SEC.TN.GEAR.DISENG", 
    "THR.BRG.ASTN.TEMP", "LO.SUMP.LVL",  "PN.AFT.SBD.BG.TEMP",
    "HMRB.PT2.PRES", "TN.GEAR.AFT.BRG.T", "LO.SUMP.TEMP", "SEC.TN.GEAR.ENGGED",
    "CLUTCH.AIR.PRES", "PN.FWD.SBD.BG.TEMP", "HMRB.PT1.PRES", 
    "LO.CLR.FW.OUT.TEMP", "LO.CLR.OUT.TEMP", "THR.BRG.OIL.TEMP", 
    "LS.GR.FWD.BRG.TEMP", "LO.HDR.PRES", "LOSP.DISCH.PRES", "LO.HDR.TEMP",
    "LO.PMP.B.DISC.PRES", "PN.AFT.PRT.BG.TEMP", "PN.FWD.PRT.BG.TEMP", 
    "LS.GR.AFT.BRG.TEMP", "LO.STR.DP") %>%
  tolower

colnames(MRG2) <- 
  c("time", "TN.GEAR.FWD.BRG.T",  "PRI.TN.GEAR.ENGGED", "PRI.TN.GEAR.DISENG", 
    "LO.PMP.A.DISC.PRES", "THR.BRG.AHEAD.TEMP", "SEC.TN.GEAR.DISENG", 
    "THR.BRG.ASTN.TEMP", "LO.SUMP.LVL",  "PN.AFT.SBD.BG.TEMP",
    "HMRB.PT2.PRES", "TN.GEAR.AFT.BRG.T", "LO.SUMP.TEMP", "SEC.TN.GEAR.ENGGED",
    "CLUTCH.AIR.PRES", "PN.FWD.SBD.BG.TEMP", "HMRB.PT1.PRES", 
    "LO.CLR.FW.OUT.TEMP", "LO.CLR.OUT.TEMP", "THR.BRG.OIL.TEMP", 
    "LS.GR.FWD.BRG.TEMP", "LO.HDR.PRES", "LOSP.DISCH.PRES", "LO.HDR.TEMP",
    "LO.PMP.B.DISC.PRES", "PN.AFT.PRT.BG.TEMP", "PN.FWD.PRT.BG.TEMP", 
    "LS.GR.AFT.BRG.TEMP", "LO.STR.DP") %>%
  tolower

names(MRG1)
names(MRG2)

MRG1$time %<>% 
  as.POSIXct(format = "%m/%d/%Y %H:%M:%S %p") %>% 
  as.numeric
MRG2$time %<>% 
  as.POSIXct(format = "%m/%d/%Y %H:%M:%S %p") %>% 
  as.numeric

write.csv(MRG1, "classC_ship1_MRG1.csv")
write.csv(MRG2, "classC_ship1_MRG2.csv")

# time <- (MRG1$time - 1473320464)/360
# MRG1$time <- time





template.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
R Template for HACKtheMACHINE September 22nd and 23rd
Created on Sun Sep 17 13:55:52 2017

@author: danielbrownjr
"""

# Import Packages
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Import package modules
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import Imputer
from sklearn.preprocessing import StandardScaler
from sklearn.cross_validation import train_test_split
from sklearn.linear_model import LinearRegression

# Import dataset
dataset = pd.read_csv('dataset.csv')
x = dataset.iloc[:, 1:-1].values
y = dataset.iloc[:, 4].values

# Taking care of missing data
imputer = Imputer(missing_values = 'NaN', strategy = 'mean', axis = 0)
imputer = imputer.fit(x[:, 1:4])
x[:, 1:4] = imputer.transform(x[:, 1:4])

# Encoding categorical data
label_encoder_x = LabelEncoder()
x[:, 0] = label_encoder_x.fit_transform(x[:, 0])
one_hot_encoder = OneHotEncoder(categorical_features = [0])
x = one_hot_encoder.fit_transform(x).toarray()

# To prevent errors in processing, in python you should use one less column
# of categorical data than the number of options in your categorical feature.
# In this case, there are two options of categorical data, so we need one
# column of features. Known as "dummy variable"
x = x[:, 1:5] # removes the "dummy variable"

# Splitting the dataset into the Training set and Test set
x_train, x_test, y_train, y_test = train_test_split(
        x, y, test_size = 0.2, random_state = 0)

# Feature Scaling
sc_x = StandardScaler()
x_train = sc_x.fit_transform(x_train)
x_test = sc_x.transform(x_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)

# Fitting Multiple Linear Regression to the Training set
regressor = LinearRegression()
regressor.fit(x_train, y_train)

# Prediction of the Test set results
y_pred = regressor.predict(x_test)

print(np.round(y_pred, 0))
print(y_test)




webDataScraper.R
#### Data Scraper Template #####################################################

### Libraries ##################################################################
install.packages("magrittr")
install.packages("tidyverse")
install.packages("rvest")
library(magrittr) # Pipe operator
library(tidyverse) # data tidying package
library(rvest) # html scraping package

### Scrape Data ################################################################
data.scraper <- rep("a", 400)
for(i in 1:400) {
  data.scraper[i] <- 
    read_html(paste0()) %>% # enter URL to be scraped
    html_node() %>% # CSS tag (Use chrome inspector)
    html_text
}

### Export Data ################################################################
write.csv(data.scraper, "scrapeddata.csv") # Exports file to .csv
data.scraper <- read.csv("scrapeddata.csv") # imports data back as dataframe


classC_ship1_MRG.R
library(tsne)
library(magrittr)
library(tidyverse)
library(cluster)

dataset <- read.csv("classC_ship1_MRG.csv")

names(dataset)

MRG1 <-
  dataset$MRG2.TN.GEAR.FWD.BRG.T %>%
  is.na %>%
  which %>%
  dataset[.,]
x1 <-
  dataset %>%
  names %>%
  grep("MRG1", .)
x1 <- c(56, x1)
MRG1 <- MRG1[, x1]

MRG2 <-
  dataset$MRG1.TN.GEAR.FWD.BRG.T %>%
  is.na %>%
  which %>%
  dataset[.,]
x2 <-
  dataset %>%
  names %>%
  grep("MRG2", .)
x2 <- c(56, x2)
MRG2 <- MRG2[, x2]

unique(MRG1.tiny)

colnames(MRG1) <- 
  c("time", "TN.GEAR.FWD.BRG.T",  "PRI.TN.GEAR.ENGGED", "PRI.TN.GEAR.DISENG", 
    "LO.PMP.A.DISC.PRES", "THR.BRG.AHEAD.TEMP", "SEC.TN.GEAR.DISENG", 
    "THR.BRG.ASTN.TEMP", "LO.SUMP.LVL",  "PN.AFT.SBD.BG.TEMP",
    "HMRB.PT2.PRES", "TN.GEAR.AFT.BRG.T", "LO.SUMP.TEMP", "SEC.TN.GEAR.ENGGED",
    "CLUTCH.AIR.PRES", "PN.FWD.SBD.BG.TEMP", "HMRB.PT1.PRES", 
    "LO.CLR.FW.OUT.TEMP", "LO.CLR.OUT.TEMP", "THR.BRG.OIL.TEMP", 
    "LS.GR.FWD.BRG.TEMP", "LO.HDR.PRES", "LOSP.DISCH.PRES", "LO.HDR.TEMP",
    "LO.PMP.B.DISC.PRES", "PN.AFT.PRT.BG.TEMP", "PN.FWD.PRT.BG.TEMP", 
    "LS.GR.AFT.BRG.TEMP", "LO.STR.DP") %>%
  tolower

colnames(MRG2) <- 
  c("time", "TN.GEAR.FWD.BRG.T",  "PRI.TN.GEAR.ENGGED", "PRI.TN.GEAR.DISENG", 
    "LO.PMP.A.DISC.PRES", "THR.BRG.AHEAD.TEMP", "SEC.TN.GEAR.DISENG", 
    "THR.BRG.ASTN.TEMP", "LO.SUMP.LVL",  "PN.AFT.SBD.BG.TEMP",
    "HMRB.PT2.PRES", "TN.GEAR.AFT.BRG.T", "LO.SUMP.TEMP", "SEC.TN.GEAR.ENGGED",
    "CLUTCH.AIR.PRES", "PN.FWD.SBD.BG.TEMP", "HMRB.PT1.PRES", 
    "LO.CLR.FW.OUT.TEMP", "LO.CLR.OUT.TEMP", "THR.BRG.OIL.TEMP", 
    "LS.GR.FWD.BRG.TEMP", "LO.HDR.PRES", "LOSP.DISCH.PRES", "LO.HDR.TEMP",
    "LO.PMP.B.DISC.PRES", "PN.AFT.PRT.BG.TEMP", "PN.FWD.PRT.BG.TEMP", 
    "LS.GR.AFT.BRG.TEMP", "LO.STR.DP") %>%
  tolower

names(MRG1)
names(MRG2)

MRG1$time %<>% 
  as.POSIXct(format = "%m/%d/%Y %H:%M:%S %p") %>% 
  as.numeric
MRG2$time %<>% 
  as.POSIXct(format = "%m/%d/%Y %H:%M:%S %p") %>% 
  as.numeric

time1 <- (MRG1$time - 1473320464)
time2 <- (MRG2$time - 1474870948)
MRG1 <- cbind(zero.time = time1, MRG1)
MRG2 <- cbind(zero.time = time2, MRG2)

x1 <- which(duplicated(MRG1))
x2 <- which(duplicated(MRG2))
MRG1.tiny <- MRG1[x1, ]
MRG2.tiny <- MRG2[x2, ]

percentage <-  (count(MRG1.tiny) + count(MRG2.tiny)) / count(dataset) * 100

plot1 <- 
  ggplot(data = MRG1.tiny) +
  aes(x = as.numeric(rownames(MRG1.tiny)), y = MRG1.tiny$zero.time) +
  geom_point(color = I("Red")) + ggtitle("Class C Ship 1 MRG 1") + xlab("Row Order") +
  ylab("POSIX time from t=0") +
  theme(axis.title.x = element_text(color = "DarkBlue", size = 16),
        axis.title.y = element_text(color = "DarkBlue", size = 16),
        plot.title = element_text(color = "DarkBlue", size = 20))

plot2 <-
  ggplot(data = MRG2.tiny) + 
  aes(x = as.numeric(rownames(MRG2.tiny)), y = MRG2.tiny$zero.time) +
  geom_point(color = I("Blue")) + ggtitle("Class C Ship 1 MRG 2") + xlab("Row Order") +
  ylab("POSIX time from t=0") +
  theme(axis.title.x = element_text(color = "DarkBlue", size = 16),
        axis.title.y = element_text(color = "DarkBlue", size = 16),
        plot.title = element_text(color = "DarkBlue", size = 20))

plot3 <-
  ggplot(data = M1A_10) + 
  aes(x = as.numeric(rownames(M1A_10)), y = M1A_10$time.zero) +
  geom_point(color = I("Purple")) + 
  ggtitle("Class C Ship 2 MPDE (sixty eight attributes)") + 
  xlab("Row Order") +
  ylab("POSIX time from t=0") +
  theme(axis.title.x = element_text(color = "DarkBlue", size = 16),
        axis.title.y = element_text(color = "DarkBlue", size = 16),
        plot.title = element_text(color = "DarkBlue", size = 20))

plot4 <-
  ggplot(data = M1A_68) + 
  aes(x = as.numeric(rownames(M1A_68)), y = M1A_68$time.zero) +
  geom_point(color = I("Orange")) + 
  ggtitle("Class C Ship 2 MPDE (ten attributes)") + 
  xlab("Row Order") +
  ylab("POSIX time from t=0") +
  theme(axis.title.x = element_text(color = "DarkBlue", size = 16),
        axis.title.y = element_text(color = "DarkBlue", size = 16),
        plot.title = element_text(color = "DarkBlue", size = 20))
multiplot(plot1, plot2, plot3, plot4)
# Multiple plot function
#
# ggplot objects can be passed in ..., or to plotlist (as a list of ggplot objects)
# - cols:   Number of columns in layout
# - layout: A matrix specifying the layout. If present, 'cols' is ignored.
#
# If the layout is something like matrix(c(1,2,3,3), nrow=2, byrow=TRUE),
# then plot 1 will go in the upper left, 2 will go in the upper right, and
# 3 will go all the way across the bottom.
#
multiplot <- function(..., plotlist=NULL, file, cols=2, layout=NULL) {
  library(grid)
  
  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)
  
  numPlots = length(plots)
  
  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                     ncol = cols, nrow = ceiling(numPlots/cols))
  }
  
  if (numPlots==1) {
    print(plots[[1]])
    
  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))
    
    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
      
      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}

#### Data Plots ###############################################################
M1A_10$DateTime.1 = NULL
time <- M1A_10$DateTime - 1473227204
M1A_10 <- cbind(time.zero = time, M1A_10)
M1A_68$DateTime.1 = NULL
time2 <- M1A_68$DateTime - 1473227204
M1A_68 <- cbind(time.zero = time2, M1A_68)

#### Unique plots #############################################################
unique.time <- unique(MRG1.tiny$zero.time)
MRG1.unique <- MRG1.tiny[unique.time,]

sort(MRG1.unique$zero.time)

plot5 <-
  ggplot(data = MRG1.unique) + 
  aes(x = as.numeric(rownames(MRG1.unique)), y = MRG1.unique$zero.time) +
  geom_point() + 
  ggtitle("Class C Ship 2 MPDE (ten attributes)") + 
  xlab("Row Order") +
  ylab("POSIX time from t=0") +
  theme(axis.title.x = element_text(color = "DarkBlue", size = 16),
        axis.title.y = element_text(color = "DarkBlue", size = 16),
        plot.title = element_text(color = "DarkBlue", size = 20))

plot5


hackthemachine_mpde.R
#### HACKtheMACHINE R Script #############################################
#
# Data for HACKtheMACHINE at MIT CSAIL Weekend Sep 22, 2017
#
# Author: Avory Bryant

library(RColorBrewer)
library("Rtsne")


# Assumes file is complete
M <- read.csv(file="classC_ship2_MPDE.csv")

M$DateTime <- as.numeric(as.POSIXlt(M$DateTime,format = "%m/%d/%Y %H:%M:%S %p"))

M$DateTime <- M$DateTime - min(M$DateTime)

timeindex <- which(names(M) == "DateTime")

fx1a <- c(timeindex,grep("X1A",names(M)))
#fx1b <- c(timeindex,grep("X1B",names(M)))
#fx2a <- c(timeindex,grep("X2A",names(M)))
#fx2b <- c(timeindex,grep("X2B",names(M)))

M1A <- M[M$Indicator == "MPDE1A",fx1a]
#M2A <- M[M$Indicator == "MPDE2A",fx2a]
#M1B <- M[M$Indicator == "MPDE1B",fx1b]
#M2B <- M[M$Indicator == "MPDE2B",fx2b]

#numna <- vector(length=nrow(M2A))
#for(i in 1:nrow(M2A)){
#	numna[i] <- sum(is.na(M2A[i,]))
#}
# print(unique(numna)) 10,68
#attr <- unique(which(!is.na(M2A[which(numna==68)[1],])))
#M2A_68 <- M2A[numna==68,c(1,attr)]
#attr <- unique(which(!is.na(M2A[which(numna==10)[1],])))
#M2A_10 <- M2A[numna==10,c(1,attr)]

# print(sum(is.na(M2A_68))) 0
# print(sum(is.na(M2A_10))) 0

numna <- vector(length=nrow(M1A))
for(i in 1:nrow(M1A)){
        numna[i] <- sum(is.na(M1A[i,]))
}
# print(unique(numna)) 10,68,45 (1)
attr <- unique(which(!is.na(M1A[which(numna==68)[1],])))
M1A_68 <- M1A[numna==68,c(1,attr)]
attr <- unique(which(!is.na(M1A[which(numna==10)[1],])))
M1A_10 <- M1A[numna==10,c(1,attr)]
attr <- unique(which(!is.na(M1A[which(numna==45)[1],])))
M1A_45 <- M1A[numna==45,c(1,attr)]
 

# print(sum(is.na(M1A_68))) 0
# print(sum(is.na(M1A_10))) 0


#numna <- vector(length=nrow(M1B))
#for(i in 1:nrow(M1B)){
#        numna[i] <- sum(is.na(M1B[i,]))
#}
# print(unique(numna)) 10 67
#attr <- unique(which(!is.na(M1B[which(numna==67)[1],])))
#M1B_67 <- M1B[numna==67,c(1,attr)]
#attr <- unique(which(!is.na(M1B[which(numna==10)[1],])))
#M1B_10 <- M1B[numna==10,c(1,attr)]
 
# print(sum(is.na(M1B_67))) 0
# print(sum(is.na(M1B_10))) 0

#numna <- vector(length=nrow(M2B))
#for(i in 1:nrow(M2B)){
#        numna[i] <- sum(is.na(M2B[i,]))
#}
# print(unique(numna)) 10 67
#attr <- unique(which(!is.na(M2B[which(numna==67)[1],])))
#M2B_67 <- M2B[numna==67,c(1,attr)]
#attr <- unique(which(!is.na(M2B[which(numna==10)[1],])))
#M2B_10 <- M2B[numna==10,c(1,attr)]
 
# print(sum(is.na(M2B_67))) 0
# print(sum(is.na(M2B_10))) 0

M1A_10_u <- unique(M1A_10)

cluster <- dbscan(matrix(M1A_10_u$DateTime,ncol=1),eps=200000,minPts=1)

tsne <- Rtsne(M1A_10_u, dims = 2, perplexity=30, verbose=TRUE, max_iter = 500)

qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
cols <- sample(length(col_vector),50)

par(mfrow=c(1,2)) 

plot(M1A_10_u$DateTime,col=cols[cluster$cluster],main="Time vs Row Order of MPDE1A-68 Depublicated Observations (Colored by Time Clustering)",ylab="Posix Time",xlab="Row Order")

plot(tsne$Y,col=cols[cluster$cluster],xlab="",ylab="",main="T-SNE 2-D Embedding of MPDE1A-68 Depublicated Observations")









hackthemachine.R
#### HACKtheMACHINE R Script #############################################
#
# Data for HACKtheMACHINE at MIT CSAIL Weekend Sep 22, 2017
#
# Author: Daniel Brown

#### Data Preprocessing ##################################################
library(magrittr)
library(tidyverse)

### Importing the dataset ################################################
dataset <- read.csv("classC_ship1_MRG.csv")
M1A_10 <- read.csv("classC_ship2_MPDE70.csv")
M1A_68 <- read.csv("classC_ship2_MPDE12.csv")

### Seperating dataset by indicator into MRG1, MRG2 ######################
MRG1 <-
  dataset$MRG2.TN.GEAR.FWD.BRG.T %>%
  is.na %>%
  which %>%
  dataset[.,]
x1 <-
  dataset %>%
  names %>%
  grep("MRG1", .)
x1 <- c(56, x1)
MRG1 <- MRG1[, x1]

MRG2 <-
  dataset$MRG1.TN.GEAR.FWD.BRG.T %>%
  is.na %>%
  which %>%
  dataset[.,]
x2 <-
  dataset %>%
  names %>%
  grep("MRG2", .)
x2 <- c(56, x2)
MRG2 <- MRG2[, x2]

### Despecifying colnames for MRG1 and MRG2 ##############################
colnames(MRG1) <- 
  c("time", "TN.GEAR.FWD.BRG.T",  "PRI.TN.GEAR.ENGGED", 
    "PRI.TN.GEAR.DISENG", "LO.PMP.A.DISC.PRES", "THR.BRG.AHEAD.TEMP", 
    "SEC.TN.GEAR.DISENG", "THR.BRG.ASTN.TEMP", "LO.SUMP.LVL",  
    "PN.AFT.SBD.BG.TEMP", "HMRB.PT2.PRES", "TN.GEAR.AFT.BRG.T", 
    "LO.SUMP.TEMP", "SEC.TN.GEAR.ENGGED", "CLUTCH.AIR.PRES", 
    "PN.FWD.SBD.BG.TEMP", "HMRB.PT1.PRES", "LO.CLR.FW.OUT.TEMP", 
    "LO.CLR.OUT.TEMP", "THR.BRG.OIL.TEMP", "LS.GR.FWD.BRG.TEMP", 
    "LO.HDR.PRES", "LOSP.DISCH.PRES", "LO.HDR.TEMP", "LO.PMP.B.DISC.PRES",
    "PN.AFT.PRT.BG.TEMP", "PN.FWD.PRT.BG.TEMP", "LS.GR.AFT.BRG.TEMP", 
    "LO.STR.DP") %>%
  tolower

colnames(MRG2) <- 
  c("time", "TN.GEAR.FWD.BRG.T",  "PRI.TN.GEAR.ENGGED", 
    "PRI.TN.GEAR.DISENG", "LO.PMP.A.DISC.PRES", "THR.BRG.AHEAD.TEMP", 
    "SEC.TN.GEAR.DISENG", "THR.BRG.ASTN.TEMP", "LO.SUMP.LVL",  
    "PN.AFT.SBD.BG.TEMP", "HMRB.PT2.PRES", "TN.GEAR.AFT.BRG.T", 
    "LO.SUMP.TEMP", "SEC.TN.GEAR.ENGGED", "CLUTCH.AIR.PRES", 
    "PN.FWD.SBD.BG.TEMP", "HMRB.PT1.PRES", "LO.CLR.FW.OUT.TEMP", 
    "LO.CLR.OUT.TEMP", "THR.BRG.OIL.TEMP", "LS.GR.FWD.BRG.TEMP", 
    "LO.HDR.PRES", "LOSP.DISCH.PRES", "LO.HDR.TEMP", "LO.PMP.B.DISC.PRES",
    "PN.AFT.PRT.BG.TEMP", "PN.FWD.PRT.BG.TEMP", "LS.GR.AFT.BRG.TEMP", 
    "LO.STR.DP") %>%
  tolower

### Converting timestamp to R recognized POSIX numeric ###################
MRG1$time %<>% 
  as.POSIXct(format = "%m/%d/%Y %H:%M:%S %p") %>% 
  as.numeric
time1 <- (MRG1$time - 1473320464)
MRG1 <- cbind(zero.time = time1, MRG1)

MRG2$time %<>% 
  as.POSIXct(format = "%m/%d/%Y %H:%M:%S %p") %>% 
  as.numeric
time2 <- (MRG2$time - 1474870948)
MRG2 <- cbind(zero.time = time2, MRG2)

### Pulling duplicated rows from MRG1 and MRG2 ###########################
x1 <- which(duplicated(MRG1))
MRG1.tiny <- MRG1[x1, ]

x2 <- which(duplicated(MRG2))
MRG2.tiny <- MRG2[x2, ]

#### Plottting Data ##########################################################

plot1 <- 
  ggplot(data = MRG1.tiny) +
  aes(x = as.numeric(rownames(MRG1.tiny)), y = MRG1.tiny$zero.time) +
  geom_point(color = I("Red")) + ggtitle("Class C Ship 1 MRG 1") + 
  xlab("Row Order") + ylab("POSIX time from t=0") +
  theme(axis.title.x = element_text(color = "DarkBlue", size = 16),
        axis.title.y = element_text(color = "DarkBlue", size = 16),
        plot.title = element_text(color = "DarkBlue", size = 20))

plot2 <-
  ggplot(data = MRG2.tiny) + 
  aes(x = as.numeric(rownames(MRG2.tiny)), y = MRG2.tiny$zero.time) +
  geom_point(color = I("Blue")) + ggtitle("Class C Ship 1 MRG 2") + 
  xlab("Row Order") + ylab("POSIX time from t=0") +
  theme(axis.title.x = element_text(color = "DarkBlue", size = 16),
        axis.title.y = element_text(color = "DarkBlue", size = 16),
        plot.title = element_text(color = "DarkBlue", size = 20))

plot3 <-
  ggplot(data = M1A_10) + 
  aes(x = as.numeric(rownames(M1A_10)), y = M1A_10$time.zero) +
  geom_point(color = I("Purple")) + 
  ggtitle("Class C Ship 2 MPDE (sixty eight attributes)") + 
  xlab("Row Order") + ylab("POSIX time from t=0") +
  theme(axis.title.x = element_text(color = "DarkBlue", size = 16),
        axis.title.y = element_text(color = "DarkBlue", size = 16),
        plot.title = element_text(color = "DarkBlue", size = 20))

plot4 <-
  ggplot(data = M1A_68) + 
  aes(x = as.numeric(rownames(M1A_68)), y = M1A_68$time.zero) +
  geom_point(color = I("Orange")) + 
  ggtitle("Class C Ship 2 MPDE (ten attributes)") + 
  xlab("Row Order") + ylab("POSIX time from t=0") +
  theme(axis.title.x = element_text(color = "DarkBlue", size = 16),
        axis.title.y = element_text(color = "DarkBlue", size = 16),
        plot.title = element_text(color = "DarkBlue", size = 20))

### Call multiplot function ##############################################
# Must load Multiplot Function from repository.
multiplot(plot1, plot2, plot3, plot4)
